### ISSUE TITLE
Internal (SPI) magnetometers occasionally fail to sample

### ISSUE THREAD

**Issue by jschall** [ArduPilot/ardupilot#5503]
#### Issue details
This is a plot over 15 hours of flight:
![internalmagproblem](https://cloud.githubusercontent.com/assets/1015761/21693635/85bc07f6-d336-11e6-96b3-3d1f508df14a.png)
The internal (SPI) magnetometers occasionally fail to sample. The external (i2c) mag does not.

#### Version
Based on dab0d8e6

#### Platform
[  ] All
[  ] AntennaTracker
[x] Copter
[  ] Plane
[  ] Rover

#### Airframe type
Quad

#### Hardware type
Pixhawk 2

**Comment by magicrub** [comment]
The airspeed sensor does this too, or at least used to until we added sanity checking spike rejection

**Comment by proficnc** [comment]
Is this using the new internal drivers? Or the PX4 drivers?

**Comment by tridge** [comment]
losing health means no reading for 500ms, which is a very long time on SPI

**Comment by tridge** [comment]
looks like this is based on fw from July 2016, so that will be the old px4 drivers. Any chance you could try again with the new driver system in master?

**Comment by OXINARF** [comment]
@tridge This is actually from November 28. GitHub loves to show the author date instead of the commit date. In any case, this is a bit old (more than 400 commits since then).

**Comment by OXINARF** [comment]
@jschall Have you seen this again? Lots of changes in drivers since then.

**Comment by jschall** [comment]
Yes, I see it on Matternet firmware, which uses the ardupilot drivers.

See matternet/ardupilot. It was rebased a couple months ago.

**Comment by tridge** [comment]
definitely worth further investigation, adding some perf counters for the various possible error conditions. I'd also suggest making the test faster by making it show unhealthy if it doesn't have a sample for 30ms or so.
I can't think of a reason why this would be happening right now, but some debug code may find it.

**Comment by peterbarker** [comment]
So 4 years on, has any further work been done on this?

@jschall did you add any logged diagnostics into matternet's code?

**Comment by IamPete1** [comment]
@jschall Bump on Peters question. 

If there are no fresh reports I this we should close.

**Comment by IamPete1** [comment]
Any updates from anyone?
